File: src/LRU_Cache.java
/*
https://leetcode.com/problems/lru-cache/

The problem can be solved with a hashtable that keeps track of the keys and its values in the double linked list. 
One interesting property about double linked list is that the node can remove itself without other reference. 
In addition, it takes constant time to add and remove nodes from the head or tail.

One particularity about the double linked list that I implemented is that I create a pseudo head and tail to mark the boundary, 
so that we don't need to check the NULL node during the update. This makes the code more concise and clean, and also it is good for the performance.
*/

import java.util.*;

public class LRU_Cache {

  class DblLinkNode {
    int key;
    int value;
    DblLinkNode prev;
    DblLinkNode post;
  }

  /**
   * Always add the new node right after head;
   */
  private void addToFront(DblLinkNode node) {
    node.prev = head;
    node.post = head.post;

    head.post.prev = node;
    head.post = node;
  }

  /**
   * Remove an existing node from the linked list.
   */
  private void removeNode(DblLinkNode node){
    DblLinkNode prev = node.prev;
    DblLinkNode post = node.post;

    prev.post = post;
    post.prev = prev;
  }

  /**
   * Move certain node in the middle to the head.
   */
  private void moveToFront(DblLinkNode node){
    this.removeNode(node);
    this.addToFront(node);
  }

  // remove the current tail. 
  private DblLinkNode removeFromEnd(){
    DblLinkNode res = tail.prev;
    this.removeNode(res);
    return res;
  }

  private Map<Integer, DblLinkNode> cache = new HashMap<>();
  private int count;
  private int capacity;
  private DblLinkNode head, tail;

  public LRU_Cache(int capacity) {
    this.count = 0;
    this.capacity = capacity;

    head = new DblLinkNode(); // dummy head
    head.prev = null;

    tail = new DblLinkNode(); // dummy tail
    tail.post = null;

    head.post = tail;
    tail.prev = head;
  }

  public int get(int key) {

    DblLinkNode node = cache.get(key);
    if(node == null){
      return -1; // should raise exception here.
    }

    // move the accessed node to the head;
    this.moveToFront(node);

    return node.value;
  }

  public void put(int key, int value) {
    DblLinkNode node = cache.get(key);

    if(node == null){

      DblLinkNode newNode = new DblLinkNode();
      newNode.key = key;
      newNode.value = value;

      this.cache.put(key, newNode);
      this.addToFront(newNode);

      ++count;

      if(count > capacity){
        // pop the tail
        DblLinkNode tail = this.removeFromEnd();
        this.cache.remove(tail.key);
        --count;
      }
    }else{
      // update the value.
      node.value = value;
      this.moveToFront(node);
    }
  }

}

File: src/C#/LRUCache.cs
/*
Design a data structure that follows the constraints of a Least Recently Used (LRU) cache.

Implement the LRUCache class:

    - LRUCache(int capacity) -- Initialize the LRU cache with positive size capacity.
    - int get(int key) -- Return the value of the key if the key exists, otherwise return -1.
    - void put(int key, int value) -- Update the value of the key if the key exists. Otherwise, add the key-value pair to the cache. If the number of keys exceeds the capacity from this operation, evict the least recently used key.

The functions get and put must each run in O(1) average time complexity.
*/

using System.Collections.Generic;

public class LRUCache {
    // Define a Node class to represent each key-value pair in the cache
    private class Node {
        public int Key;
        public int Value;
        public Node Prev;
        public Node Next;
    }

    /*
    The idea is to use a Doubly Linked List to keep track of the least recently used and most recently used nodes. The least recently used node is always at the tail of the list, and the most recently used node is always at the head of the list. We also use a dictionary to look up nodes by their corresponding keys.
    */
    private readonly int capacity;
    private readonly Dictionary<int, Node> cache;
    private Node head; // most recently used node
    private Node tail; // least recently used node

    public LRUCache(int capacity) {
        this.capacity = capacity;
        cache = new Dictionary<int, Node>(capacity);
        head = null;
        tail = null;
    }

    public int Get(int key) {
        // If key exists in cache, move its corresponding node to the front of the list and return its value
        if (cache.TryGetValue(key, out Node node)) {
            MoveToFront(node);
            return node.Value;
        }
        // If key does not exist in cache, return -1
        return -1;
    }

    public void Put(int key, int value) {
        if (cache.TryGetValue(key, out Node node)) {
            // If key already exists in cache, update its value and move its corresponding node to the front of the list
            node.Value = value;
            MoveToFront(node);
        } else {
            // If key does not exist in cache, add its corresponding node to the front of the list
            node = new Node { Key = key, Value = value };
            cache.Add(key, node);
            if (cache.Count > capacity) {
                // If cache capacity is exceeded, remove the least recently used node from the cache and the list
                cache.Remove(tail.Key);
                RemoveFromTail();
            }
            AddToFront(node);
        }
    }

    private void MoveToFront(Node node) {
        if (node == head) {
            return;
        }
        if (node == tail) {
            tail = node.Prev;
        } else {
            node.Next.Prev = node.Prev;
        }
        node.Prev.Next = node.Next;
        AddToFront(node);
    }

    private void AddToFront(Node node) {
        if (head == null) {
            head = node;
            tail = node;
        } else {
            node.Next = head;
            head.Prev = node;
            head = node;
        }
    }

    private void RemoveFromTail() {
        if (tail == null) {
            return;
        }
        if (tail == head) {
            head = null;
            tail = null;
        } else {
            tail.Prev.Next = null;
            tail = tail.Prev;
        }
    }
}
